src/script_opt/ZAM/ZBody.cc

```diff
@@ -0,0 +1,539 @@
+
+	if ( zc->NonRecursive() )
+		{
+		fixed_frame = new ZVal[frame_size];
+
+		for ( auto& ms : managed_slots )
+			fixed_frame[ms].ClearManagedVal();
+		}
+
+	table_iters = zc->GetTableIters();
+	num_step_iters = zc->NumStepIters();
+
+	// It's a little weird doing this in the constructor, but unless
+	// we add a general "initialize for ZAM" function, this is as good
+	// a place as any.
+	if ( ! did_init )
+		{
+		auto log_ID_type = lookup_ID("ID", "Log");
+		ASSERT(log_ID_type);
+		log_ID_enum_type = log_ID_type->GetType<EnumType>();
+
+		any_base_type = base_type(TYPE_ANY);
+
+		ZVal::SetZValNilStatusAddr(&ZAM_error);
+
+		did_init = false;
+		}
+	}
+
+ZBody::~ZBody()
+	{
+	delete[] fixed_frame;
+	delete[] insts;
+	delete inst_count;
+	delete CPU_time;
+	}
+
+void ZBody::SetInsts(vector<ZInst*>& _insts)
+	{
+	ninst = _insts.size();
+	auto insts_copy = new ZInst[ninst];
+
+	for ( auto i = 0U; i < ninst; ++i )
+		insts_copy[i] = *_insts[i];
+
+	insts = insts_copy;
+
+	InitProfile();
+	}
+
+void ZBody::SetInsts(vector<ZInstI*>& instsI)
+	{
+	ninst = instsI.size();
+	auto insts_copy = new ZInst[ninst];
+
+	for ( auto i = 0U; i < ninst; ++i )
+		{
+		auto& iI = *instsI[i];
+		insts_copy[i] = iI;
+		if ( iI.stmt )
+			insts_copy[i].loc = iI.stmt->Original()->GetLocationInfo();
+		}
+
+	insts = insts_copy;
+
+	InitProfile();
+	}
+
+void ZBody::InitProfile()
+	{
+	if ( analysis_options.profile_ZAM )
+		{
+		inst_count = new vector<int>;
+		inst_CPU = new vector<double>;
+		for ( auto i = 0U; i < ninst; ++i )
+			{
+			inst_count->push_back(0);
+			inst_CPU->push_back(0.0);
+			}
+
+		CPU_time = new double;
+		*CPU_time = 0.0;
+		}
+	}
+
+ValPtr ZBody::Exec(Frame* f, StmtFlowType& flow)
+	{
+#ifdef DEBUG
+	double t = analysis_options.profile_ZAM ? curr_CPU_time() : 0.0;
+#endif
+
+	auto val = DoExec(f, 0, flow);
+
+#ifdef DEBUG
+	if ( analysis_options.profile_ZAM )
+		*CPU_time += curr_CPU_time() - t;
+#endif
+
+	return val;
+	}
+
+ValPtr ZBody::DoExec(Frame* f, int start_pc, StmtFlowType& flow)
+	{
+	int pc = start_pc;
+	const int end_pc = ninst;
+
+	// Return value, or nil if none.
+	const ZVal* ret_u;
+
+	// Type of the return value.  If nil, then we don't have a value.
+	TypePtr ret_type;
+
+#ifdef DEBUG
+	bool do_profile = analysis_options.profile_ZAM;
+#endif
+
+	ZVal* frame;
+	std::unique_ptr<TableIterVec> local_table_iters;
+	std::vector<StepIterInfo> step_iters(num_step_iters);
+
+	if ( fixed_frame )
+		frame = fixed_frame;
+	else
+		{
+		frame = new ZVal[frame_size];
+		// Clear slots for which we do explicit memory management.
+		for ( auto s : managed_slots )
+			frame[s].ClearManagedVal();
+
+		if ( ! table_iters.empty() )
+			{
+			local_table_iters =
+			    std::make_unique<TableIterVec>(table_iters.size());
+			*local_table_iters = table_iters;
+			tiv_ptr = &(*local_table_iters);
+			}
+		}
+
+	flow = FLOW_RETURN;	// can be over-written by a Hook-Break
+
+	while ( pc < end_pc && ! ZAM_error ) {
```

Comment:
nit: The style guide says this curly brace should be on the next line.

Reply:
Fixed.